### PR TITLE
[boost-ext-mp] Add new port

### LINKED ISSUE #29269: [New Port Request] boost-ext-mp
Library name: boost-ext-mp

Library description: C++ single header/single module C++20 Meta-Programming Library

Source repository URL: https://github.com/boost-ext/mp

Project homepage (if different from the source repository): https://boost-ext.github.io/mp/

Anything else that is useful to know when adding (such as optional features the library may have that should be included): The boost-ext-mp library support using `import boost.mp` to import the library, it will be fantastic if the vcpkg can support the modules.

### PR DESCRIPTION
**Pull request by ex-purple**
Fixes https://github.com/microsoft/vcpkg/issues/29269

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by MonicaLiu0311** [comment]
usage:
```
boost-ext-mp provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(mp CONFIG REQUIRED)
    target_link_libraries(main PRIVATE Boost::mp)
```

When testing usage, the following error occurs:
```
>------ Build All started: Project: CMakeFindUsage, Configuration: x64-Release ------
  MSBuild version 17.4.1+9a89d02ff for .NET Framework
    Checking Build System
    Building Custom Rule C:/Users/monica/source/repos/CMakeFindUsage/CMakeFindUsage/CMakeLists.txt
    CMakeFindUsage.cpp
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(171,13): error C3527: 'Cs' is not a valid operand for 'sizeof...'. Did you mean to use 'sizeof'? 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(569,24): error C3861: '__PRETTY_FUNCTION__': identifier not found 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(569,24): error C2065: '__PRETTY_FUNCTION__': undeclared identifier 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(590,8): error C2065: '__PRETTY_FUNCTION__': undeclared identifier 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(590,8): error C3861: '__PRETTY_FUNCTION__': identifier not found 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(591,14): error C2065: '__PRETTY_FUNCTION__': undeclared identifier 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(591,14): error C3861: '__PRETTY_FUNCTION__': identifier not found 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(597,8): error C2065: '__PRETTY_FUNCTION__': undeclared identifier 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(597,8): error C3861: '__PRETTY_FUNCTION__': identifier not found 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(598,14): error C2065: '__PRETTY_FUNCTION__': undeclared identifier 
E:\boost-ext-mp\installed\x64-windows\include\mp-0.0.1\boost\mp.hpp(598,14): error C3861: '__PRETTY_FUNCTION__': identifier not found 

Build All failed.
```

<details>
<summary><b>CMakeFindUsage.cpp</b></summary>
<pre>
#include &#60;iostream&#62;
#include "boost/mp.hpp"

using namespace std;

int main()
{
        cout << "Hello CMake." << endl;
        return 0;
}
</pre></details>


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>
cmake_minimum_required (VERSION 3.8)

set(CMAKE_TOOLCHAIN_FILE "E:/boost-ext-mp/scripts/buildsystems/vcpkg.cmake")

project ("CMakeFindUsage")

find_package(mp CONFIG REQUIRED)

add_executable (CMakeFindUsage "CMakeFindUsage.cpp")

target_link_libraries(CMakeFindUsage PRIVATE Boost::mp)
</pre></details>

**Comment by ex-purple** [comment]
@MonicaLiu0311 
MSVC doesn't support __PRETTY_FUNCTION__. I just can disable win and uwp platforms for this port

**Comment by Neumann-A** [comment]
Clang-cl probably does support it

**Comment by ex-purple** [comment]
I have no idea what else I can do with this port

**Comment by dg0yt** [comment]
There are  `__FUNCDNAME__` and `__FUNCSIG__` in Visual C++, but I don't know if they provide what is needed in this particular case. https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170
This is something to discuss upstream.

**Comment by MonicaLiu0311** [comment]
Waiting for the reply from the upstream for the time being.

https://github.com/boost-ext/mp/issues/65

File: ports/boost-ext-mp/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "boost-ext-mp",
+  "version-date": "2023-03-02",
+  "description": "C++20 ~~Template~~ Meta-Programming",
+  "homepage": "https://boost-ext.github.io/mp/",
+  "license": "BSL-1.0",
+  "supports": "!(win | uwp)",
```
Comment (MonicaLiu0311) [inline ports/boost-ext-mp/vcpkg.json:7]: By the way: Your expression does not work for Windows, it should be: "supports": "!(windows | uwp)"

Details: [vcpkg-json.md#-platform-expression](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/reference/vcpkg-json.md#-platform-expression)

Comment (dg0yt) [reply ports/boost-ext-mp/vcpkg.json:7]: It should be `!windows | mingw`.  Windows includes uwp and mingw, but mingw uses gcc or clang and will probably work (untested).

Comment (ex-purple) [reply ports/boost-ext-mp/vcpkg.json]: > It should be !windows | mingw.
Looks good. I agree with it

**Comment by MonicaLiu0311** [comment]
There is no response from the upstream, so we will test it on linux. The following are the test results:


* usage(x64-linux):
```
boost-ext-mp provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(mp CONFIG REQUIRED)
    target_link_libraries(main PRIVATE Boost::mp)
```

* When testing usage, the following error occurs:
```
monica:~/CMakeFindUsage/build$ cmake --build .
[ 50%] Building CXX object CMakeFiles/main.dir/CMakeFindUsage.cpp.o
In file included from /home/monica/CMakeFindUsage/CMakeFindUsage.cpp:4:
/home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:7: error: expected identifier before ‘typeof’
  208 | using typeof = decltype(expr(std::declval<Ts>()...));
      |       ^~~~~~
/home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:14: error: expected primary-expression before ‘=’ token
  208 | using typeof = decltype(expr(std::declval<Ts>()...));
      |              ^
/home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:14: error: expected unqualified-id before ‘=’ token
gmake[2]: *** [CMakeFiles/main.dir/build.make:76: CMakeFiles/main.dir/CMakeFindUsage.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/main.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
```

<details>
<summary><b>CMakeFindUsage.cpp</b></summary>
<pre>
#include &#60;iostream&#62;
#include "boost/mp.hpp"

using namespace std;

int main()
{
        cout << "Hello CMake." << endl;
        return 0;
}
</pre></details>


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>
cmake_minimum_required (VERSION 3.8)

set(CMAKE_TOOLCHAIN_FILE "/home/monica/mp/scripts/buildsystems/vcpkg.cmake")

project ("CMakeFindUsage")

find_package(mp CONFIG REQUIRED)

add_executable (main "CMakeFindUsage.cpp")

target_link_libraries(main PRIVATE Boost::mp)
</pre></details>

**Comment by ex-purple** [comment]
@MonicaLiu0311 
> "description": "C++20 ~~Template~~ Meta-Programming",

So you have to install the appropriate C++ standard in the CMakeLists.txt:
`set(CMAKE_CXX_STANDARD 20)`

**Comment by MonicaLiu0311** [comment]
> Waiting for the reply from the upstream for the time being.
> 
> [boost-ext/mp#65](https://github.com/boost-ext/mp/issues/65)

No response from upstream, so continue with this PR.

**Comment by MonicaLiu0311** [comment]
> @MonicaLiu0311
> 
> > "description": "C++20 ~Template~ Meta-Programming",
> 
> So you have to set the appropriate C++ standard in the CMakeLists.txt: `set(CMAKE_CXX_STANDARD 20)`

I still have this error after setting `set(CMAKE_CXX_STANDARD 20)` on `linux`.

**Comment by jbbjarnason** [comment]
Is there any progress with this port?

**Comment by MonicaLiu0311** [comment]
I have clearly specified `set(CMAKE_CXX_STANDARD 20)`, but when testing Usage, the following error still occurs:
```
root@monica003:/home/monica/test/build# cmake --build .
[ 50%] Building CXX object CMakeFiles/main.dir/test.cpp.o
In file included from /home/monica/test/test.cpp:3:
/home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:7: error: expected identifier before ‘typeof’
  208 | using typeof = decltype(expr(std::declval<Ts>()...));
      |       ^~~~~~
/home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:14: error: expected primary-expression before ‘=’ token
  208 | using typeof = decltype(expr(std::declval<Ts>()...));
      |              ^
/home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:14: error: expected unqualified-id before ‘=’ token
gmake[2]: *** [CMakeFiles/main.dir/build.make:76: CMakeFiles/main.dir/test.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/main.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
```

<details>
<summary><b>CMakeFindUsage.cpp</b></summary>
<pre>
#include &#60;iostream&#62;
#include "boost/mp.hpp"

using namespace std;

int main()
{
        cout << "Hello CMake." << endl;
        return 0;
}
</pre></details>


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>
cmake_minimum_required (VERSION 3.8)

set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_STANDARD 20)

set(CMAKE_TOOLCHAIN_FILE "E:/mp/scripts/buildsystems/vcpkg.cmake")

project ("CMakeFindUsage")

add_executable (main "CMakeFindUsage.cpp")

find_package(mp CONFIG REQUIRED)
target_link_libraries(main PRIVATE Boost::mp)

</pre></details>


[CMakeOutput.log](https://github.com/microsoft/vcpkg/files/11411717/CMakeOutput.log)
[CMakeCache.txt](https://github.com/microsoft/vcpkg/files/11411718/CMakeCache.txt)

**Comment by ex-purple** [comment]
@MonicaLiu0311
I wonder that compiler with C++20 support cannot build that.
In any case, it's not a port issue. It's just a boost-ext-mp header that the compiler is either able to understand or not.

**Comment by MonicaLiu0311** [comment]
The error is pointing to the following piece of code:
```
template <auto expr, class... Ts>
using typeof = decltype(expr(std::declval<Ts>()...));
```
When I replace the `typeof` keyword with standard C++'s `typeid` operator, the error disappears:
```
//template <auto expr, class... Ts>
//using typeof = decltype(expr(std::declval<Ts>()...));
template <typename type, typename... Ts>
using decltype_type = decltype(typeid(type(std::declval<Ts>()...)));
```


And the boost used upstream is outdated, and the new version of the boost library has removed the use of the `typeof` keyword.

**Comment by ex-purple** [comment]
@MonicaLiu0311 
> When I replace the typeof keyword with standard C++'s typeid operator, the error disappears:

> And the boost used upstream is outdated, and the new version of the boost library has removed the use of the typeof keyword.

Looks like this should be fixed by upstream, not here.

**Comment by MonicaLiu0311** [comment]
> Looks like this should be fixed by upstream, not here.

Reported upstream: https://github.com/boost-ext/mp/issues/68

Thanks for your contribution. Yes, I understand that there may be upstream issues causing this, but we need to ensure that the port added to vcpkg is usable. Would you be willing to try this approach(https://github.com/microsoft/vcpkg/pull/29975#issuecomment-1539795139) to see if it passes on your machine, and then patch it to fix it?

**Comment by ex-purple** [comment]
> then patch it to fix it

I propose to transfer this PR to a draft or archive. Because patching the source code is, in my opinion, a bad idea.
It's one thing when a patch fixes problems with building or installation, and another thing when the source code of the project changes

**Comment by MonicaLiu0311** [comment]
The upstream has not responded for more than 3 months, so this PR is closed now. If the upstream solves issues https://github.com/boost-ext/mp/issues/65 and https://github.com/boost-ext/mp/issues/68 later, please open it again.

**Comment by jiayuehua** [comment]
> I have clearly specified `set(CMAKE_CXX_STANDARD 20)`, but when testing Usage, the following error still occurs:
> 
> ```
> root@monica003:/home/monica/test/build# cmake --build .
> [ 50%] Building CXX object CMakeFiles/main.dir/test.cpp.o
> In file included from /home/monica/test/test.cpp:3:
> /home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:7: error: expected identifier before ‘typeof’
>   208 | using typeof = decltype(expr(std::declval<Ts>()...));
>       |       ^~~~~~
> /home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:14: error: expected primary-expression before ‘=’ token
>   208 | using typeof = decltype(expr(std::declval<Ts>()...));
>       |              ^
> /home/monica/mp/installed/x64-linux/include/mp-0.0.1/boost/mp.hpp:208:14: error: expected unqualified-id before ‘=’ token
> gmake[2]: *** [CMakeFiles/main.dir/build.make:76: CMakeFiles/main.dir/test.cpp.o] Error 1
> gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/main.dir/all] Error 2
> gmake: *** [Makefile:91: all] Error 2
> ```
> 
> CMakeFindUsage.cpp
> CMakeLists.txt
> [CMakeOutput.log](https://github.com/microsoft/vcpkg/files/11411717/CMakeOutput.log) [CMakeCache.txt](https://github.com/microsoft/vcpkg/files/11411718/CMakeCache.txt)

After some invertigation , I now know why. Your main target in CMakeList.txt need add property like this :

 `set_property(TARGET main PROPERTY  CXX_EXTENSIONS  OFF)`

This change build flags from `-std=gnu++20` to `-std=c++20` and then the main target can build successfully. I have reopen the pr.

**Pull request closed without merging**